ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoiceNotifier.java

```diff
@@ -90,8 +94,68 @@ public void onAttestationsDue(final UInt64 slot) {
     eventThread.execute(() -> internalAttestationsDue(slot));
   }
 
-  public SafeFuture<Optional<Bytes8>> getPayloadId() {
-    return eventThread.execute(() -> lastPayloadId);
+  public SafeFuture<Optional<Bytes8>> getPayloadId(Bytes32 beaconBlockRoot, UInt64 slot) {
+    return eventThread.execute(() -> internalGetPayloadId(beaconBlockRoot, slot));
+  }
+
+  public void onTTDReached(Bytes32 executionBlockHash) {
+    eventThread.execute(() -> internalTTDReached(executionBlockHash));
+  }
+
+  private void internalTTDReached(Bytes32 executionBlockHash) {
+    eventThread.checkOnEventThread();
+    executionPayloadTerminalBlockHash = Optional.of(executionBlockHash);
+  }
+
+  /**
+   * must return Optional.empty() only when is safe to produce a block with an empty execution
+   * payload Optional.of(payloadId) when it is certain that belongs to the given beacon block root
+   * and slot
+   *
+   * <p>in all other cases it must Throw to avoid block production
+   */
+  private Optional<Bytes8> internalGetPayloadId(Bytes32 beaconBlockRoot, UInt64 slot) {
```

Comment:
I think I'd rename `beaconBlockRoot` to `beaconBlockParentRoot` to be clear that this is the root of the parent of the block we're trying to create.